test/image/mocks/zz-pie_pattern.json

```diff
@@ -0,0 +1,89 @@
+{
+  "data": [
+    {
+      "labels": [ "giraffe", "orangutan", "monkey"],
+      "values": [ 30, 20, 50],
+      "type": "pie",
+      "marker": {
+        "colors": [ "red", "green", "red"],
+        "line": {
+          "color": "lightgrey",
+          "width": 4
+        },
+        "pattern": {
+          "shape": [ "+", "", "-"],
+          "fgcolor": [ "darksalmon", "black", "steelblue"],
+          "bgcolor": "lightgrey"
+        }
+      },
+      "textfont": { "color": ["black", "white", "black"]},
```

Comment:
I wonder if we can get this to be the default result for inside text? This would particularly help in the cases where we don't know whether the text will end up inside or outside. Maybe if the pattern solidity <0.5 we contrast with the bgcolor, >0.5 we contrast with the fgcolor? Here's where this auto color is chosen (`Color.contrast` picks either black or white, whichever is farther from the given color):
https://github.com/plotly/plotly.js/blob/0919988d818cc04d6c093539638dde05c998cc73/src/traces/pie/plot.js#L531

Note I haven't looked but wonder if we want to do the same for bars: 

https://github.com/plotly/plotly.js/blob/0919988d818cc04d6c093539638dde05c998cc73/src/traces/bar/style.js#L124

Reply:
I checked similar cases for bar trace and scatter trace, see attached. Scatter trace with area filled with a pattern seems a special case anyway, but also the actual bar trace is deciding on the color of the text inside the bar by evaluating the base color of the bar, without taking into account pattern attributes.

And even flipping the text color to black inside the bar / slice will give a distortion effect, probably as bad as the actual situation. Patterns with text inside is probably never a very good idea ..

I would prefer to add a pattern to a few other trace types first (funnelarea, sunburst, ..) and to handle the color of the text inside a pattern eventually later via a separate PR.
![zz-pattern_textfont_color](https://github.com/plotly/plotly.js/assets/11560767/34a99375-292f-4b32-a9d8-ff5e54ca07d5)
[zz-pattern_textfont_color.json.txt](https://github.com/plotly/plotly.js/files/11565611/zz-pattern_textfont_color.json.txt)

Reply:
Hmm OK - I guess it's not so clear that any color we choose here will actually be usable, for a lot of situations neither black nor white would really work and (given that this is often used in black&white contexts) we don't want to consider any other colors. So the recommendation I guess would be to use `textposition='outside'` when you have patterns.